Documentation added for testing

razajafri opened the pull request:
This PR starts the process of improving our documentation for running integration and unit-tests 

This addresses #814  
<!--

Thank you for contributing to RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

razajafri commented:
@sameerz PTAL

jlowe commented on CONTRIBUTING.md:
```diff
@@ -28,7 +28,7 @@ Contributions to RAPIDS Accelerator for Apache Spark fall into the following thr
     labels.
 3. Comment on the issue stating that you are going to work on it.
-4. Code! Make sure to update unit tests!
+4. Code! Make sure to update unit tests and integration tests if needed! [refer to test section]
```
Suggest making the "refer to test section" a hyperlink to the section.

jlowe commented on CONTRIBUTING.md:
```diff
@@ -56,6 +56,35 @@ This project follows the
 and the Scala conventions detailed above, preferring the latter.
 
+### Test
+There are two types of tests in this project. Unit tests that are written in Scala, and integration tests that are written in python. 
```
```suggestion
There are two types of tests in this project. Unit tests that are written in Scala, and integration tests that are written in Python. 
```

jlowe commented on CONTRIBUTING.md:
```diff
@@ -56,6 +56,35 @@ This project follows the
+We encourage writing integration tests if you have to choose between the two as that helps us ensure the plugin is working across different platforms. 
+
+#### Unit tests
+Unit-tests are located in com.nvidia:rapids-4-spark-tests_2.12. In order to run the unit-tests follow these steps
```
I don't think listing a maven group and artifact ID is helpful here.  Why not point them to the test Maven project in this very repo?

jlowe commented on CONTRIBUTING.md:
```diff
@@ -56,6 +56,35 @@ This project follows the
+
+#### Unit tests
+Unit-tests are located in com.nvidia:rapids-4-spark-tests_2.12. In order to run the unit-tests follow these steps
+1. Build the parent project. That will build the entire project including tests
```
Similarly I think parent project is more nebulous than it needs to be here.  "Build from the top of the repository" would be clearer, IMHO.

jlowe commented on CONTRIBUTING.md:
```diff
@@ -56,6 +56,35 @@ This project follows the
+Unit-tests are located in com.nvidia:rapids-4-spark-tests_2.12. In order to run the unit-tests follow these steps
+1. Build the parent project. That will build the entire project including tests
+2. Running tests
+    2a. Issue the mvn command to run the tests like 'mvn -pl com.nvidia:rapids-4-spark-tests_2.12 test'. This will run all the tests in the tests module. 
```
Why not the much more straightforward and expected `mvn test` here?  I'd rather avoid directing users to invoking specific maven projects, and it's standard to expect `mvn test` to run all the unit tests as it does today.

jlowe commented on CONTRIBUTING.md:
```diff
@@ -56,6 +56,35 @@ This project follows the
+There are two types of tests in this project. Unit tests that are written in Scala, and integration tests that are written in python. 
+We encourage writing integration tests if you have to choose between the two as that helps us ensure the plugin is working across different platforms. 
+
+#### Unit tests
```
The bullets in this section are not formatted properly.  See the Markdown rendering at https://github.com/NVIDIA/spark-rapids/blob/a90286e980aa74eff0191d2e496dd8ee6832724b/CONTRIBUTING.md.

In addition the bullet points are quite wordy.  If there's that much to say per bullet, I'd suggest using sub-headings with useful section header names rather than bullets.

jlowe commented on CONTRIBUTING.md:
```diff
@@ -56,6 +56,35 @@ This project follows the
+2. Running tests
+    2a. Issue the mvn command to run the tests like 'mvn -pl com.nvidia:rapids-4-spark-tests_2.12 test'. This will run all the tests in the tests module. 
+    2b. To run individual tests append '-DwildcardSuites=<Fully-qualified-test-name>' to the above command e.g. To run test ParquetWriterSuite in package com.nvidia.spark.rapids, issue 'mvn test -DwildcardSuites="com.nvidia.spark.rapids.ParquetWriterSuite"'
+3. Running tests against different Spark versions. There are maven profiles that should be used to run Spark version-specific tests e.g. To run tests against Spark 3.1.0, 'mvn -P spark310tests test -DwildcardSuites="com.nvidia.spark.rapids.ParquetWriterSuite"'. Please refer to the pom.xml to see the list of profiles supported
```
```suggestion
3. Running tests against different Apache Spark versions. There are Maven profiles that should be used to run Spark version-specific tests e.g. To run tests against Spark 3.1.0, `mvn -P spark310tests test -DwildcardSuites=com.nvidia.spark.rapids.ParquetWriterSuite`. Please refer to the pom.xml to see the list of profiles supported
```

jlowe commented on CONTRIBUTING.md:
```diff
@@ -56,6 +56,35 @@ This project follows the
+1. Build the parent project. That will build the entire project including tests
+2. Running tests
+    2a. Issue the mvn command to run the tests like 'mvn -pl com.nvidia:rapids-4-spark-tests_2.12 test'. This will run all the tests in the tests module. 
+    2b. To run individual tests append '-DwildcardSuites=<Fully-qualified-test-name>' to the above command e.g. To run test ParquetWriterSuite in package com.nvidia.spark.rapids, issue 'mvn test -DwildcardSuites="com.nvidia.spark.rapids.ParquetWriterSuite"'
```
Use backticks when quoting code or things the user should type verbatim.  This applies in multiple places.

jlowe commented on CONTRIBUTING.md:
```diff
@@ -56,6 +56,35 @@ This project follows the
+'SPARK_CONF="spark.rapids.memory.gpu.allocFraction=0.7,spark.sql.cache.serializer=com.nvidia.spark.rapids.shims.spark310.ParquetCachedBatchSerializer" mvn ...' this command sets the memory-allocation fraction which comes handy when dealing with OOM issues on the GPU.
+
+#### Integration tests
+Integration tests are located in com.nvidia:rapids-4-spark-integration-tests_2.12. The suggested way to run these tests is to use the shell-script file located in the module folder called 'run_pyspark_from_build.sh'. This script takes care of some of the flags that are required to run the tests e.g. `--conf spark.driver.extraJavaOptions=-Duser.timezone=GMT --conf spark.sql.session.timeZone=UTC -- ...
```
This should not refer to the Maven artifact, especially since that doesn't help the user find the scripts.  Hyperlink to the repository directory and/or the shell scripts.

Also recommend removing the details like the flags being passed, this documentation is likely to become obsolete/wrong in those areas and arguably isn't interesting/relevant at this level of documentation.

`run_pyspak_from_build.sh` is mis-spelled.

jlowe replied:
Please word-wrap the lines being added to this file so it's friendlier to read in terminals and easier to comment on in PRs.

jlowe commented on CONTRIBUTING.md:
```diff
@@ -56,6 +56,35 @@ This project follows the
+The tests are written python and run with pytest and the script honors pytest parameters. Some handy flags are:
+    - `-k` <pytest-file-name>. This will run all the tests in that test file.
+    - `-k` <test-name>. This will also run an individual test.
+    - `-s`. Doesn't capture the output and instead prints to the screen.
```
Suggest adding `-v -rfExXs` to get more details from each test run.

jlowe commented on CONTRIBUTING.md:
```diff
@@ -56,6 +56,35 @@ This project follows the
+    - For other options and more details please visit [pytest-usage](https://docs.pytest.org/en/stable/usage.html) or type `pytest --help`
+
+Example: 
+This command runs all the tests located in `cache_test.py` against Spark-3.1.0 using the ParquetCachedBatchSerializer and other conigs discussed above and with the debugger listening on port 5005
```
`conigs` is misspelled.

jlowe commented on CONTRIBUTING.md:
```diff
@@ -56,6 +56,35 @@ This project follows the
 [Oracle Java code conventions](http://www.oracle.com/technetwork/java/codeconvtoc-136057.html)
 and the Scala conventions detailed above, preferring the latter.
 
+### Test
```
```suggestion
### Testing Your Code
```

I also suggest moving this after `Sign Your Work` since that's missed by most new devs to the project and being closer to the top of the file would help.

There's enough stuff in here about testing that we may want to move it into the docs/dev directory and just link to it from this contributing document under a Testing Your Code header.

razajafri commented:
build

GaryShen2008 commented:
build

razajafri commented:
build

tgravescs commented on docs/dev/testing.md:
```diff
@@ -0,0 +1,44 @@
+- To run test ParquetWriterSuite in package com.nvidia.spark.rapids, issue `mvn test -DwildcardSuites="com.nvidia.spark.rapids.ParquetWriterSuite"`
+
+
+#### Integration tests
```
so there is already documentation in integration_tests.md do we really need this in both places?  or just have a link and put the details in that doc

revans2 replied:
I would prefer to have us link to it.  There is also documentation about unit tests 

https://github.com/NVIDIA/spark-rapids/blob/branch-0.3/docs/testing.md

Not sure why we are not linking to that primarily and updating it as needed.

razajafri replied:
I will link to the testing doc and update if needed. I honestly wasn't aware we have a document already, thanks for bringing it up

tgravescs replied:
I'm still confused, why do we have so many different docs for testing. I realize you linked them but now there is info in testing.md and integrationtest.md . We added information here on running but I need to go to integration tests readme.md to know what i need to install to add it.  why don't we put all information in the integration test readme and test readme and just have small description and point to those?   Maybe i'm missing something @revans2 @jlowe opinions?

tgravescs commented on docs/dev/testing.md:
```diff
@@ -0,0 +1,44 @@
+better insight into what is needed as we constantly keep working on to improve and expand the plugin-support.
+
+The tests are written python and run with pytest and the script honors pytest parameters. Some handy flags are:
+- `-k` <pytest-file-name>. This will run all the tests in that test file.
```
I guess I haven't tried this, usually I specify the path to the test file to run all tests in a file. I supposed both works? or is this specific to the run_pyspark_from_build.sh script? I have been doing spark-submit way.

revans2 replied:
`-k` does a test search for the name of the test.  This includes the file name and the extensions for specific parameters.  I prefer -k because I can do things like `-k Decimal` and I will typically end up running all of the tests that touch the decimal type either as a parameter or something else.

But it would be good to document both ways.

tgravescs commented on docs/dev/testing.md:
```diff
@@ -0,0 +1,44 @@
+- `-k` <pytest-file-name>. This will run all the tests in that test file.
+- `-k` <test-name>. This will also run an individual test.
+- `-s` Doesn't capture the output and instead prints to the screen.
+- `-v` increase verbosity of the tests
```
perhaps add the debug tmp path option that save the output

razajafri replied:
are you referring to `--debug`? This option doesn't accept path. Are you referring to some other option?

revans2 commented on docs/dev/testing.md:
```diff
@@ -0,0 +1,44 @@
+### Testing your code
+There are two types of tests in this project. Unit tests that are written in Scala, and integration tests that are written in Python. 
```
We actually have integration tests in both Scala and python.

razajafri commented:
build

razajafri commented:
build

razajafri commented:
@tgravescs @revans2 I have addressed your concerns PTAL

revans2 commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
 
-Unit tests exist in the tests directory. This is unconventional and is done so we can run the tests
-on the final shaded version of the plugin. It also helps with how we collect code coverage.
+Unit tests exist in the [here](tests) directory. This is unconventional and is done so we can run
```
"unit tests exist in the here directory" is not very readable.  Also not sure why you are linking to.  There is no tests directory or file under docs that this can or should link to.

revans2 commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
+Unit tests exist in the [here](tests) directory. This is unconventional and is done so we can run
+ the tests on the final shaded version of the plugin. It also helps with how we collect code coverage.
+
+in order to run the unit-tests follow these steps
```
Please capitalize sentences.

revans2 commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
+
+in order to run the unit-tests follow these steps
+1. issue the maven command to run the tests with `mvn test`. this will run all the tests
+2. to run individual tests append `-dwildcardsuites=<comma separated list of wildcard suite names to execute>` to the above command 
```
These are not steps.  They are options and I don't think we need a list to express them.

revans2 commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
+1. issue the maven command to run the tests with `mvn test`. this will run all the tests
+2. to run individual tests append `-dwildcardsuites=<comma separated list of wildcard suite names to execute>` to the above command 
+
+for more information about using scalatest with maven please refere [here](https://www.scalatest.org/user_guide/using_the_scalatest_maven_plugin)
```
I'm not known for my spelling but I don't think refer has an "e" at the end

revans2 commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
-There are two frameworks used for testing. One is based off of pytest and pyspark in the 
-`src/main/python` directory. These tests will run as a part of the maven build if you have the environment
-variable `SPARK_HOME` set.
+Integration tests are stored in the [integration_tests](integration_tests) directory. 
```
This is not in the right directory for this link to work.  Please change it back to the README or at least ../integration_tests

revans2 commented on docs/testing.md:
````diff
@@ -117,3 +151,12 @@ Next you can start to run the tests.
 ...
 ```
+
+For more details on Integration tests options please refer to integration-test [doc](integration-tests/README.md)
````
Need to test out the links I don't think this works.

razajafri commented:
@revans2 better now? The links were all copied from where I added this section originally in the CONTRIBUTING.md. 

Thanks for reviewing

razajafri commented:
build

tgravescs commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
+
+In order to run the unit-tests issue the maven command to run the tests i.e. `mvn test`. 
+
+To run individual tests append `-DwildcardSuites=<comma separated list of wildcard suite names to execute>` to the above command 
```
does this work for both scala and java tests? I know the java ones are different in spark

jlowe replied:
wildcardSuites only applies to scalatest.  Java tests need to be filtered via surefire's approach which is `-Dtest=...`

tgravescs replied:
right so we have a java test right - TestHashedPriorityQueue.java?  so adding how to run it would be useful and specifying here this works for scala tests.

tgravescs replied:
I guess this text is old new one mentions scala tests

jlowe commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
+- `-k` <test-name>. This will also run an individual test.
+- `-s` Doesn't capture the output and instead prints to the screen.
+- `-v` Increase the verbosity of the tests
+- `-rfexxs` Show extra test summary info as specified by chars: (f)ailed, (e)rror, (x)failed, (x)passed, (s)kipped
```
These flags are wrong.  The first clue is the same character meaning two different options.  It should be something like this:
```suggestion
- `-r fExXs` Show extra test summary info as specified by chars: (f)ailed, (E)rror, (x)failed, (X)passed, (s)kipped
```

jlowe commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
-Unit tests exist in the tests directory. This is unconventional and is done so we can run the tests
-on the final shaded version of the plugin. It also helps with how we collect code coverage.
+Unit tests exist in the [tests](../tests) directory. This is unconventional and is done so we can
+ run the tests on the final shaded version of the plugin. It also helps with how we collect code coverage.
```
```suggestion
run the tests on the final shaded version of the plugin. It also helps with how we collect code coverage.
```

jlowe commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
+Unit tests exist in the [tests](../tests) directory. This is unconventional and is done so we can
+ run the tests on the final shaded version of the plugin. It also helps with how we collect code coverage.
+
+In order to run the unit-tests issue the maven command to run the tests i.e. `mvn test`. 
```
```suggestion
Use Maven to run the unit tests via `mvn test`. 
```

jlowe commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
+
+To run individual tests append `-DwildcardSuites=<comma separated list of wildcard suite names to execute>` to the above command 
+
+For more information about using scalatest with maven please refer [here](https://www.scalatest.org/user_guide/using_the_scalatest_maven_plugin)
```
```suggestion
For more information about using scalatest with Maven please refer to the
[scalatest documentation](https://www.scalatest.org/user_guide/using_the_scalatest_maven_plugin).
```

jlowe commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
+   - `-Pspark302tests` (spark 3.0.2)
+   - `-Pspark310tests` (spark 3.1.0)
+
+Please refer to the [pom.xml](tests/pom.xml) to see the list of profiles supported
+apache spark specific configurations can be passed in by setting environment-variable spark_conf 
```
```suggestion
Please refer to the [tests project POM](tests/pom.xml) to see the list of test profiles supported.
Apache Spark specific configurations can be passed in by setting the `SPARK_CONF` environment
variable.
```

jlowe commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
+apache spark specific configurations can be passed in by setting environment-variable spark_conf 
+
+Examples: 
+-To run tests against Apache Spark 3.1.0, 
```
This says it will run tests against Spark 3.1.0 but only runs one test.  Either fix the description or fix the example.

jlowe commented on docs/testing.md:
```diff
@@ -39,20 +39,54 @@ They generally follow TPCH but are not guaranteed to be the same.
-variable `SPARK_HOME` set.
+Integration tests are stored in the [integration_tests](../integration_tests) directory. 
+There are two frameworks used for testing. one is based off of pytest and pyspark in the 
+`src/main/python` directory. These tests will run as a part of the maven build if you have the 
```
```suggestion
`src/main/python` directory. These tests will run as a part of the Maven build if you have the 
```

jlowe commented:
> why don't we put all information in the integration test readme and test readme and just have small description and point to those?

I'm all for not repeating information across the various docs and instead linking.  I agree we should keep unit test info to the unit test docs and integration test info to the integration test docs.  Each doc can reference the other for more information on running the unit tests vs. integration tests.

revans2 commented:
> I'm all for not repeating information across the various docs and instead linking. I agree we should keep unit test info to the unit test docs and integration test info to the integration test docs. Each doc can reference the other for more information on running the unit tests vs. integration tests.

I'm not a big fan of pointing things from the docs directory outside of it, simply because the docs directory gets moved around when it becomes the github pages site.  But because this is docs for developers I think @jlowe is right and we should have it all in the README for the various directories with links to them, so long as we are careful to make sure that the links work on the github pages site and in markdown.

razajafri commented:
@revans2 @tgravescs @jlowe 

Sorry about the back and forth. I have incorporated all the suggestions I think. Please check if this looks good. I have also added a section on the project README.md for tests and linked it to tests/README.md. Let me know if that looks good

razajafri commented:
build

razajafri commented:
build

razajafri commented:
@revans2 @tgravescs do you have any other suggestions or does this look OK to go in?

jlowe commented on CONTRIBUTING.md:
```diff
@@ -28,7 +28,7 @@ Contributions to RAPIDS Accelerator for Apache Spark fall into the following thr
     labels.
 3. Comment on the issue stating that you are going to work on it.
-4. Code! Make sure to update unit tests!
+4. Code! Make sure to update unit tests and integration tests if needed! [refer to test section](#test)
```
```suggestion
4. Code! Make sure to update unit tests and integration tests if needed! [refer to test section](#testing-your-code)
```

jlowe commented on CONTRIBUTING.md:
````diff
@@ -119,6 +119,8 @@ By making a contribution to this project, I certify that:
     this project or the open source license(s) involved.
 ```
 
+### Testing your code
````
```suggestion
### Testing Your Code
```

jlowe commented on tests/README.md:
```diff
@@ -0,0 +1,74 @@
+For more information about using scalatest with Maven please refer to the
+[scalatest documentation](https://www.scalatest.org/user_guide/using_the_scalatest_maven_plugin).
+    
+#### Running unit-tests against specific apache spark versions. 
```
```suggestion
#### Running Unit Tests Against Specific Apache Spark Versions
```

jlowe commented on tests/README.md:
```diff
@@ -0,0 +1,74 @@
+They generally follow TPCH but are not guaranteed to be the same.
+`Q1Like(spark)` will return a DataFrame that can be executed to run the corresponding query.
+
+## Unit tests
```
```suggestion
## Unit Tests
```

jlowe commented on tests/README.md:
```diff
@@ -0,0 +1,74 @@
+ `SPARK_CONF="spark.dynamicAllocation.enabled=false,spark.task.cpus=1" mvn ...`
+- To run test ParquetWriterSuite in package com.nvidia.spark.rapids, issue `mvn test -DwildcardSuites="com.nvidia.spark.rapids.ParquetWriterSuite"`
+
+## Integration tests
```
```suggestion
## Integration Tests
```

jlowe commented on tests/README.md:
```diff
@@ -0,0 +1,74 @@
+---
+# RAPIDS Accelerator for Apache Spark Testing
+
+We have several stand alone examples that you can run in the integration_tests subdirectory/jar.
```
```suggestion
We have several stand alone examples that you can run in the [integration tests](../integration_tests).
```

razajafri commented:
build

razajafri commented:
@jlowe thanks for the suggestions. They have all been applied. Please take another look

jlowe reviewed (approved):
This looks OK to me, but @revans2 and @tgravescs should comment before merging.

tgravescs reviewed (approved):
overall is fine, I think we have one java unit test that we could say how to run individually but not a big deal

razajafri commented:
Thanks, @tgravescs I have created a follow-on #1345
